File: C#/s/s/Exp.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace s
{
    public class Exp
    {
        public enum ExpType
        {
            Exp_Large,
            Exp_Medium,
            Exp_Small,
            Exp_String,
            Exp_Int,
            Exp_Bool,
            Exp_Id,/*id包含kvs-path*/
            Exp_Let,
            Exp_LetId,
            Exp_LetSmall,
            Exp_LetRest
        }
        private bool isBracket;
        public bool IsBracket()
        {
            return isBracket;
        }
        private ExpType type;
        public ExpType Exp_type()
        {
            return type;
        }
        /*括号部分*/
        private Node<Exp> children;
        private Node<Exp> r_children;
        public Node<Exp> Children() { return children; }
        public Node<Exp> R_children() { return r_children; }
        private Token left;
        private Token right;
        public Token Left() { return left; }
        public Token Right() { return right; }
        private Exp(
            ExpType type, 
            Token left,
            Token right,
            Node<Exp> children, 
            Node<Exp> r_children
        )
        {
            this.isBracket = true;
            this.type = type;
            this.left = left;
            this.right = right;
            this.children = children;
            this.r_children = r_children;
        }

        /*基本组*/
        private Token token;
        public Token Token() { return token; }
        public String Value() { return token.Value(); }
        private int int_value;
        public int Int_Value(){ return int_value;}
        private bool bool_value;
        public bool Bool_Value() { return bool_value; }
        private Node<String> kvs_paths;
        public Node<String> KVS_paths() { return kvs_paths; }
        public Exp(ExpType type, Token token)
        {
            this.isBracket = false;
            this.type = type;
            this.token = token;
            if (type == ExpType.Exp_Int)
            {
                int_value = int.Parse(token.Value());
            }
            if (type == ExpType.Exp_Bool)
            {
                bool_value=(token.Value() == "true");
            }
            if (type == ExpType.Exp_Id)
            {
                String id = token.Value();
                if (id[0] == '.' || id[id.Length - 1] == '.')
                {
                    //throw new LocationException(loc,"不是合法的id类型，不能以.开始或结束");
                    kvs_paths = null;
                }
                else
                {
                    int i = 0;
                    int last_i = 0;
                    Node<String> r = null;
                    bool has_error = false;
                    while (i < id.Length)
                    {
                        char c = id[i];
                        if (c == '.')
                        {
                            String node = id.Substring(last_i, i - last_i);
                            last_i = i + 1;
                            if (node == "")
                            {
                                has_error = true;
                            }
                            else
                            {
                                r = Node<String>.extend(node, r);
                            }
                        }
                        i++;
                    }
                    /*最后一个，默认匹配*/
                    r = Node<String>.extend(id.Substring(last_i), r);
                    if (has_error)
                    {
                        throw this.exception(id + "不是合法的id类型，不允许连续的.号");/*id类型中间有连续的..*/
                    }
                    else
                    {
                        kvs_paths = Node<String>.reverse(r);
                    }
                }
            }
        }
        public void toString(StringBuilder sb)
        {
            if (isBracket)
            {
                sb.Append(left.Value());
                for (Node<Exp> t = children; t != null; t = t.Rest())
                {
                    sb.Append(t.First().ToString()).Append(" ");
                }
                sb.Append(right.Value());
            }
            else
            {
                sb.Append(token.ToString());
            }
        }
        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            toString(sb);
            return sb.ToString();
        }
        public LocationException exception(String msg)
        {
            if (isBracket)
            {
                return new LocationException(left.Loc(), msg);
            }
            else
            {
                return new LocationException(token.Loc(), msg);
            }
        }
        public void warn(String msg)
        {
            if (isBracket)
            {
                Console.Write(left.Loc().ToString());
            }
            else
            {
                Console.Write(token.Loc().ToString());
            }
            Console.WriteLine(msg);
        }

        /// <summary>
        /// 重设Let中的ID表达式。
        /// </summary>
        /// <param name="k"></param>
        /// <returns></returns>
        static Exp resetLetID(Exp k)
        {
            if (k.Value().IndexOf('.') < 0)
            {
                return new Exp(ExpType.Exp_LetId, k.Token());
            }
            else
            {
                throw k.exception("let表达式中，" + k.ToString() + "不是合法的key-id类型");/*let-id中间不允许有任何的.*/
            }
        }
        /// <summary>
        /// let表达式的，括号匹配的部分，不过用其反转的子来计算
        /// </summary>
        /// <param name="k"></param>
        /// <returns></returns>
        static Exp resetLetSmall(Exp small)
        {
            Node<Exp> vs = small.R_children();
            Node<Exp> children=null;
            if (vs != null)
            {
                Exp k = vs.First();
                vs = vs.Rest();
                if (k.Exp_type() == ExpType.Exp_Id)
                {
                    String v = k.Value();
                    if (v.StartsWith("..."))
                    {
                        v = v.Substring(3);
                        if (v.IndexOf('.') < 0)
                        {
                            children = Node<Exp>.extend(
                                new Exp(
                                    ExpType.Exp_LetRest,
                                    new Token(v,k.Token().Old_Value(),k.Token().Token_type(),k.Token().Loc())
                                ),
                                children
                             );
                        }
                        else
                        {
                            throw k.exception("let表达式中，" + k.ToString() + "不是合法的剩余匹配ID");/*剩余匹配的id，只允许三个开头的点*/
                        }
                    }
                    else
                    {
                        /*
                         * 最后一个只是普通的ID
                         */
                        children = Node<Exp>.extend(resetLetID(k),children);
                    }
                }
            }
            while (vs != null)
            {
                Exp k = vs.First();
                vs = vs.Rest();
                if (k.Exp_type() == ExpType.Exp_Small)
                {
                    children = Node<Exp>.extend(resetLetSmall(k), children);
                }
                else if (k.Exp_type() == ExpType.Exp_Id)
                {
                    children = Node<Exp>.extend(resetLetID(k),children);
                }
                else
                {
                    throw k.exception("Let表达式中，不是合法的key类型" + k.ToString());/*Let表达式中，key部分出现其它类型（如数字、布尔等）不应该允许出现*/
                }
            }
            return new Exp(
                ExpType.Exp_LetSmall,
                small.Left(),
                small.Right(),
                children,
                null
            );
        }

        /// <summary>
        /// let表达式kvkv的部分，不过是反转的，即vkvk...
        /// </summary>
        /// <param name="vks"></param>
        /// <returns></returns>
        static Node<Exp> resetLetVKS(Node<Exp> vks)
        {
            Node<Exp> children = null;
            while (vks != null)
            {
                Exp v = vks.First();
                children = Node<Exp>.extend(v,children);
                vks = vks.Rest();
                if (vks != null)
                {
                    Exp k = vks.First();
                    vks = vks.Rest();

                    if (k.Exp_type() == ExpType.Exp_Id)
                    {
                        children=Node<Exp>.extend(resetLetID(k),children);
                    }
                    else if (k.Exp_type() == ExpType.Exp_Small)
                    {
                        if (k.Children() == null)
                        {
                            k.warn("Let表达式中无意义的空()，请检查:" + v.ToString());/*let表达式中，无意义的空*/
                        }
                        children=Node<Exp>.extend(resetLetSmall(k),children);
                    }
                    else
                    {
                        throw k.exception("let表达式中，不合法的key类型:" + k.ToString());/*let表达式中，不是合法的key类型，不是id或()*/
                    }
                }
                else
                {
                    throw v.exception("let表达式中期待与value:" + v.ToString() + "匹配，却结束了let表达式");/*let表达式不完全*/
                }
            }
            return children;
        }
        /// <summary>
        /// 检查函数内的无用函数
        /// </summary>
        /// <param name="children"></param>
        static void check_Large(Node<Exp> vs)
        {
            while (vs != null)
            {
                Exp v = vs.First();
                vs = vs.Rest();
                if (vs != null)
                {
                    ExpType t = v.Exp_type();
                    if(!(t==ExpType.Exp_Let || t==ExpType.Exp_Small || t==ExpType.Exp_Medium))
                    {
                        v.warn("函数中定义无意义的表达式，请检查:"+v.ToString());
                    }
                }
            }
        }
        public static Exp Parse(Node<Token> tokens)
        {
            Location root_loc = new Location(0, 0, 0);
            Token root_left = new Token("{", "{", s.Token.TokenType.Token_BracketLeft, root_loc);
            Token root_right = new Token("}", "}", s.Token.TokenType.Token_BracketRight, root_loc);
            Exp exp = new Exp(ExpType.Exp_Large, root_left, root_right, null, null);
            Node<Exp> caches = Node<Exp>.extend(exp, null);
            Node<Token> xs = tokens;
            Node<Exp> children = null;
            while (xs != null)
            {
                Token x = xs.First();
                xs = xs.Rest();
                if (x.Token_type() == s.Token.TokenType.Token_BracketRight)
                {
                    Exp.ExpType tp;
                    if (x.Value() == ")")
                    {
                        tp = ExpType.Exp_Small;
                    }
                    else if (x.Value() == "]")
                    {
                        tp = ExpType.Exp_Medium;
                    }
                    else
                    {
                        tp = ExpType.Exp_Large;
                    }
                    caches = Node<Exp>.extend(
                        new Exp(tp, null, x, children, null),
                        caches
                    );
                    children = null;
                }
                else if (x.Token_type() == s.Token.TokenType.Token_BracketLeft)
                {
                    Exp cache = caches.First();
                    Node<Exp> r_children = null;
                    ExpType tp = cache.Exp_type();

                    if (tp == ExpType.Exp_Large)
                    {
                        check_Large(children);
                    }
                    else
                    {
                        r_children = Node<Exp>.reverse(children);
                    }
                    Node<Exp> caches_parent = caches.Rest();
                    if (caches_parent != null)
                    {
                        Exp p_exp = caches_parent.First();
                        if (p_exp.Exp_type() == ExpType.Exp_Large)
                        {
                            //父表达式是函数
                            if (tp == ExpType.Exp_Small)
                            {
                                if (children == null)
                                {
                                    throw new LocationException(x.Loc(),"不允许空的()");/*左括号*/
                                }
                                else
                                {
                                    Exp first = children.First();
                                    if (first.Exp_type() == ExpType.Exp_Id && first.Value() == "let")
                                    {
                                        tp = ExpType.Exp_Let;
                                        if (children.Length() == 1)
                                        {
                                            throw first.exception("不允许空的let表达式");/*let标识的位置*/
                                        }
                                        else
                                        {
                                            children = Node<Exp>.extend(children.First(), resetLetVKS(Node<Exp>.reverse(children.Rest())));
                                        }
                                    }
                                    else
                                    {
                                        if (!(first.Exp_type() == ExpType.Exp_Id || first.Exp_type() == ExpType.Exp_Large || first.Exp_type() == ExpType.Exp_Small))
                                        {
                                            throw first.exception("函数调用第一个应该是id或{}或()，而不是" + first.ToString());/*first已经计算出来了*/
                                        }
                                    }
                                }
                            }
                        }
                    }

                    children =Node<Exp>.extend(
                        new Exp(
                            tp,
                            x,
                            cache.Right(),
                            children,
                            r_children
                        ),
                        cache.Children()
                    );
                    caches = caches_parent;
                }
                else
                {
                    ExpType tp=0;
                    bool deal = true;
                    if (x.Token_type() == s.Token.TokenType.Token_String)
                    {
                        tp = ExpType.Exp_String;
                    }
                    else if (x.Token_type() == s.Token.TokenType.Token_Int)
                    {
                        tp = ExpType.Exp_Int;
                    }
                    else if (x.Token_type() == s.Token.TokenType.Token_Bool)
                    {
                        tp = ExpType.Exp_Bool;
                    }
                    else
                    {
                        Exp parent = caches.First();
                        if (parent.Exp_type() == ExpType.Exp_Medium)
                        {
                            if (x.Token_type() == s.Token.TokenType.Token_Prevent)
                            {
                                if (x.Value() == "true" || x.Value() == "false" || s.Token.isInt(x.Value()))
                                {
                                    throw new LocationException(x.Loc(), "中括号内转义寻找作用域上的" + x.Value() + "定义");/*token相关*/
                                }
                                tp = ExpType.Exp_Id;
                            }
                            else if (x.Token_type() == s.Token.TokenType.Token_Id)
                            {
                                tp = ExpType.Exp_String;
                            }
                            else
                            {
                                deal = false;
                            }
                        }
                        else
                        {
                            if (x.Token_type() == s.Token.TokenType.Token_Prevent)
                            {
                                tp = ExpType.Exp_String;
                            }
                            else if (x.Token_type() == s.Token.TokenType.Token_Id)
                            {
                                tp = ExpType.Exp_Id;
                            }
                            else
                            {
                                deal = false;
                            }
                        }
                    }

                    if (deal)
                    {
                        children =Node<Exp>.extend(
                            new Exp(tp, x),
                            children
                        );
                    }
                }
            }
            check_Large(children);
            return new Exp(ExpType.Exp_Large, root_left,root_right,children, null);
        }
    }
}

File: Python/s/Exp.py
# coding=utf-8

from Token import Token,TokenType
from Location import Location
from Node import Node
from LocationException import LocationException

class ExpType:
    Exp_Large=0
    Exp_Medium=1
    Exp_Small=2
    Exp_String=3
    Exp_Int=4
    Exp_Bool=5
    Exp_Id=6
    Exp_Let=7
    Exp_LetId=8
    Exp_LetSmall=9
    Exp_LetRest=10
    
class Exp:
    """docstring for Exp"""
    def __init__(self, type):
        self.exp_type=type
        
    def exception(self,msg):
        pass
    def warn(self,msg):
        pass
        
    def isBracket(self):
        pass
    
    @staticmethod
    def resetLetID(k):
        if k.value.find('.')<0:
            return AtomExp(ExpType.Exp_LetId,k.token)
        else:
            raise k.exception("let表达式中，"+str(k)+"不是合法的key-id类型")
    
    @staticmethod    
    def resetLetSmall(small):
        vs=small.r_children
        children=None
        if vs!=None:
            k=vs.First()
            vs=vs.Rest()
            if k.exp_type==ExpType.Exp_Id:
                v=k.value
                if v.find("...")==0:
                    v=v[3:len(v)]
                    if v.find('.')<0:
                        children=Node.extend(
                            AtomExp(
                                ExpType.Exp_LetRest,
                                Token(v,k.token.old_value,k.token.token_type,k.token.loc)
                            ),
                            children
                        )
                    else:
                        raise k.exception("let表达式中，"+str(k)+"不是合法的剩余匹配ID")
                else:
                    children=Node.extend(Exp.resetLetID(k),children)
            while vs!=None:
                k=vs.First()
                vs=vs.Rest()
                if k.exp_type==ExpType.Exp_Small:
                    children=Node.extend(Exp.resetLetSmall(k),children)
                elif k.exp_type==ExpType.Exp_Id:
                    children=Node.extend(Exp.resetLetID(k),children)
                else:
                    raise k.exception("Let表达式中，不是合法的key类型"+str(k))
        return BracketExp(
            ExpType.Exp_LetSmall,
            small.left,
            children,
            small.right
        )
    
    @staticmethod
    def resetLetVKS(vks):
        children=None
        while vks!=None:
            v=vks.First()
            children=Node.extend(v,children)
            vks=vks.Rest()
            if vks!=None:
                k=vks.First()
                vks=vks.Rest()
                if k.exp_type==ExpType.Exp_Id:
                    children=Node.extend(Exp.resetLetID(k),children)
                elif k.exp_type==ExpType.Exp_Small:
                    if k.children==None:
                        raise k.warn("Let表达式中无意义 的空()，请检查："+str(v))
                    children=Node.extend(Exp.resetLetSmall(k),children)
                else:
                    raise k.exception("let表达式中，不合法的key类型:"+str(k))
            else:
                raise v.exception("let表达式中期待与value:"+str(v)+"匹配，却结束了let表达式")
        return children
    
    @staticmethod
    def check_Large(vs):
        while vs!=None:
            v=vs.First()
            vs=vs.Rest()
            if vs!=None:
                t=v.exp_type
                if (not (t==ExpType.Exp_Let or t==ExpType.Exp_Small or t==ExpType.Exp_Medium)):
                    v.warn("函数中定义无意义的表达式，请检查"+str(v))
    
    @staticmethod
    def Parse(tokens):
        root_loc=Location(0,0,0)
        root_left=Token("{","{",TokenType.Token_BracketLeft,root_loc)
        root_right=Token("}","}",TokenType.Token_BracketRight,root_loc)
        exp=BracketExp(ExpType.Exp_Large,root_left,None,root_right,None)
        caches=Node.extend(exp,None)
        xs=tokens
        children=None
        while xs!=None:
            x=xs.First()
            xs=xs.Rest()
            if x.token_type==TokenType.Token_BracketRight:
                tp=None
                if x.value==")":
                    tp=ExpType.Exp_Small
                elif x.value=="]":
                    tp=ExpType.Exp_Medium
                elif x.value=="}":
                    tp=ExpType.Exp_Large
                else:
                    raise LocationException(x.loc,"不正常的左匹配")
                caches=Node.extend(
                    BracketExp(tp,None,children,x),
                    caches
                )
                children=None
            elif x.token_type==TokenType.Token_BracketLeft:
                cache=caches.First()
                r_children=None
                tp=cache.exp_type
                if tp==ExpType.Exp_Large:
                    Exp.check_Large(children)
                else:
                    r_children=Node.reverse(children)
                caches_parent=caches.Rest()
                if caches_parent!=None:
                    p_exp=caches_parent.First()
                    if p_exp.exp_type==ExpType.Exp_Large:
                        #父表达式为函数
                        if tp==ExpType.Exp_Small:
                            if children==None:
                                raise LocationException(x.loc,"不允许空的()")
                            else:
                                first=children.First()
                                if (first.exp_type==ExpType.Exp_Id and first.value=="let"):
                                    tp=ExpType.Exp_Let
                                    if children.Length()==1:
                                        raise first.exception("不允许空的let表达式")
                                    else:
                                        children=Node.extend(
                                            children.First(),
                                            Exp.resetLetVKS(Node.reverse(children.Rest()))
                                        )
                                else:
                                    if (not (first.exp_type==ExpType.Exp_Large or first.exp_type==ExpType.Exp_Id or first.exp_type==ExpType.Exp_Small)):
                                        raise first.exception("函数调用第一个应该是id或{}或()，而不是"+str(first.exp_type)+str(first))
                children=Node.extend(
                    BracketExp(
                        tp,
                        x,
                        children,
                        cache.right,
                        r_children
                    ),
                    cache.children
                )
                caches=caches_parent
            else:
                tp=None
                deal=True
                if x.token_type==TokenType.Token_String:
                    tp=ExpType.Exp_String
                elif x.token_type==TokenType.Token_Int:
                    tp=ExpType.Exp_Int
                elif x.token_type==TokenType.Token_Bool:
                    tp=ExpType.Exp_Bool
                else:
                    parent=caches.First()
                    if parent.exp_type==ExpType.Exp_Medium:
                        if x.token_type==TokenType.Token_Prevent:
                            if(x.value=="true" or x.value=="false" or Token.isInt(x.value)):
                                raise LocationException(x.loc,"中括号中转义寻找作用域上的"+x.value)
                            tp=ExpType.Exp_Id
                        elif x.token_type==TokenType.Token_Id:
                            tp=ExpType.Exp_String
                        else:
                            deal=False
                    else:
                        if x.token_type==TokenType.Token_Prevent:
                            tp=ExpType.Exp_String
                        elif x.token_type==TokenType.Token_Id:
                            tp=ExpType.Exp_Id
                        else:
                            deal=False
                            
                if deal:
                    e=None
                    if tp==ExpType.Exp_Id:
                        e=IDExp(x)
                    elif tp==ExpType.Exp_Int:
                        e=IntExp(x)
                    elif tp==ExpType.Exp_Bool:
                        e=BoolExp(x)
                    else:
                        e=AtomExp(tp,x)
                    children=Node.extend(e,children)
        Exp.check_Large(children)
        return BracketExp(ExpType.Exp_Large,root_left,children,root_right)
class AtomExp(Exp):
    def __init__(self,type,token):
        Exp.__init__(self, type)
        self.token=token
        self.value=token.value
        
    def exception(self, msg):
        return LocationException(self.token.loc,msg)
    def warn(self, msg):
        print(str(self.token.loc)+msg)
        
    def isBracket(self):
        return False
    
    def __str__(self):
        return self.token.old_value
    
class IntExp(AtomExp):
    def __init__(self,token):
        AtomExp.__init__(self,ExpType.Exp_Int,token)
        self.int_value=int(token.value)
        
class BoolExp(AtomExp):
    def __init__(self,token):
        AtomExp.__init__(self,ExpType.Exp_Bool,token)
        self.bool_value=True if token.value=="true"  else False
        
class IDExp(AtomExp):
    def __init__(self,token):
        AtomExp.__init__(self,ExpType.Exp_Id,token)
        v=token.value
        len_v=len(v)
        if (v[0]=='.' or v[len_v-1]=='.'):
           self.paths=None
        else:
            i=0
            last_i=0
            r=None
            has_error=False
            while i<len_v:
                c=v[i]
                if c=='.':
                    node=v[last_i:i]
                    last_i=i+1
                    if node=="":
                        has_error=True
                    else:
                        r=Node.extend(node,r)
                i=i+1
            r=Node.extend(v[last_i:len_v],r)
            if has_error:
                raise LocationException(token.loc,v+"不是合法的ID，不允许连续的。")
            else:
                self.paths=Node.reverse(r)
                
class BracketExp(Exp):
    def __init__(self,type,left,children,right,r_children=None):
        Exp.__init__(self, type)
        self.left=left
        self.children=children
        self.right=right
        self.r_children=r_children
        
    def warn(self, msg):
        print(msg,self.left.loc,self.right.loc)
        
    def exception(self, msg):
        return LocationException(self.left.loc,msg)
    
    def isBracket(self):
        return True
    
    def __str__(self):
        sb=[];
        sb.append(self.left.old_value)
        t=self.children
        while t!=None:
            sb.append(str(t.First()))
            sb.append(" ")
            t=t.Rest()
        sb.append(self.right.old_value)
        return "".join(sb)

'''
tokens=Token.run("99 (let a 76 (b (c d)...e) []) (log 98 [a b {} \"abcde\" c d e])", "\n")
print(tokens)
try:
    exps=Exp.Parse(tokens)
    print(exps)
except LocationException,lex:
    print str(lex)
'''